### PR TITLE
S2 local minimum information

### PR DESCRIPTION
**Pull request by JYangQi00**
A plugin for looking at the S2 local minimum information. This plugin provides the depth and width of the local minimum's valley, as well as the max goodness of split of the S2 peak after merging. Hopefully this information can be used to do some cuts for multiple scatters.

_Before you submit this PR: make sure to put all operations-related information in a wiki-note, a PR should be about code and is publicly accessible_

## What does the code in this PR do / what does it improve?
This PR adds a LoopPlugin over the main S2 peaks and provides information about the S2 local minimum (if it has one). The purpose of this is to look for which S2s are bi-modal peaks, as this may be due to a double scatter that happens close in x-y but is delayed in z.
## Can you briefly describe how it works?
This plugin smooths the main S2 peak and provides the max goodness of split, the depth of the local minimum, and the width of the local minimum. These three parameters can be studied as figures of merit in order to make multiple-scatter cuts for those peaks which don't show two distinct populations in their hitpattern, but do show a hint at two separate peaks in their time signature.
## Can you give a minimal working example (or illustrate with a figure)?
Please see the last section of this note:

https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenonnt:jianyang:s2_split_tagging

_Please include the following if applicable:_
  - [ ] _Update the docstring(s)_
  - [ ] _Update the documentation_
  - [ ] _Tests to check the (new) code is working as desired._
  - [ ] _Does it solve one of the open issues on github?_

### _Notes on testing_
 - _Until the automated tests pass, please mark the PR as a draft._
 - _On the XENONnT fork we test with database access, on private forks there is no database access for security considerations._

All _italic_ comments can be removed from this template.

### CODE REVIEW

**Comment by skazama** [comment]
@JYangQi00 Is this PR still up-to-date?

**Comment by JYangQi00** [comment]
Hi Shingo, sorry about the late reply. I updated it and tested it out (merged with the latest master branch of straxen) and it works and is up to date.

**Comment by JoranAngevaare** [comment]
Hi @JYangQi00 looks like a nice plugin! 

Can you make sure the code is tested and do you maybe have some numbers on performance, how long does it take to process a typical run?

Maybe also keep in mind that the options names propagate to the entire context, should we rename them a bit such that they aren't so general, e.g. `smoothing_power` could mean many things if you see it somewhere popping up in the context config?

File: straxen/plugins/local_minimum_info.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    :param smoothp: The smoothed peak
+    :param max_loc: Location of every local maximum of the peak
+    :param min_loc: Location of every local minimum of the peak
+    :param pv: "Percent value", the percent of the valley height for which to calculate the gap
+    :param dt: The time of one sample in ns
+    :return: The gap in ns, the depth in PE
+    """
+    n_gap = len(min_loc)
+    p_length = len(smoothp)
+    if (len(max_loc) - n_gap) == 1 and len(min_loc):
```
Comment (WenzDaniel) [inline straxen/plugins/local_minimum_info.py:124]: You could invert this statement and directly return 0,0. In this way you can remove the additional else and indentation.

**Review by WenzDaniel — COMMENTED** [review]
Hej @JYangQi00, the PR looks already quite nice and clean. I have not worked through logic of the code though. I think it would be good if we could add some unit tests to the function you wrote. This will help us to spot possible bugs or bugs when making changes in the future.

**Comment by JYangQi00** [comment]
> Hi @JYangQi00 looks like a nice plugin!
> 
> Can you make sure the code is tested and do you maybe have some numbers on performance, how long does it take to process a typical run?
> 
> Maybe also keep in mind that the options names propagate to the entire context, should we rename them a bit such that they aren't so general, e.g. `smoothing_power` could mean many things if you see it somewhere popping up in the context config?

Hi Joran, thanks for your comments. I changed the config options to have a "_localmin" after them to avoid the confusion. I was unable to test this on a background run since I am using a local directory. However, I tested it on the Kr83 run "018161" and after all the dependent data types were made, it took 2 minutes and 13 seconds to process event_local_min_info.

**Comment by JYangQi00** [comment]
> Hej @JYangQi00, the PR looks already quite nice and clean. I have not worked through logic of the code though. I think it would be good if we could add some unit tests to the function you wrote. This will help us to spot possible bugs or bugs when making changes in the future.

Hi Daniel, thanks for your suggestions. I made the appropriate adjustments as suggested. It took 2 minutes 13 seconds to process the run "018161". Unfortunately, I am using a local installation of straxen to test out this branch, so I can't load background runs, the one I processed was a Kr83m run.

**Comment by JoranAngevaare** [comment]
> > Hi @JYangQi00 looks like a nice plugin!
> > Can you make sure the code is tested and do you maybe have some numbers on performance, how long does it take to process a typical run?
> > Maybe also keep in mind that the options names propagate to the entire context, should we rename them a bit such that they aren't so general, e.g. `smoothing_power` could mean many things if you see it somewhere popping up in the context config?
> 
> Hi Joran, thanks for your comments. I changed the config options to have a "_localmin" after them to avoid the confusion. I was unable to test this on a background run since I am using a local directory. However, I tested it on the Kr83 run "018161" and after all the dependent data types were made, it took 2 minutes and 13 seconds to process event_local_min_info.

Thanks @JYangQi00! Sorry, I meant if you could also write a unit-test for some of the functions you wrote (see for examples the `tests/` folder on github.

Thanks for checking out the speed as well. 2 minutes is more than I'd have expected for a high level plugin. Do you know where the bottleneck resides? With some [profiling](https://pypi.org/project/line-profiler/) you might be able to tell. What also might help is writing it as an ordinary plugin (like `EventBasics`) such that you can use numpy vectorization and/or numba more.

**Pull request closed without merging**